Add ability for _WKNodeInfo to be used as an argument with WKWebView.callAsyncJavaScript

achristensen07 opened the pull request:
#### 2dbebd6180a99f5c3e6f282707eac2fde1482b8a
<pre>
Add ability for _WKNodeInfo to be used as an argument with WKWebView.callAsyncJavaScript
<a href="https://bugs.webkit.org/show_bug.cgi?id=296836">https://bugs.webkit.org/show_bug.cgi?id=296836</a>
<a href="https://rdar.apple.com/157338015">rdar://157338015</a>

Reviewed by Ryosuke Niwa.

If the node does not exist any more, I chose to pass JS undefined in as a parameter.
This makes _WKNodeInfo kind of like a WeakRef, which returns undefined if the node doesn&apos;t exist any more.
If the node exists in a different document, I decided to also return JS undefined.
Both of these choices are arbitrary and non-final at this point.  Still prototyping its use.

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toJS):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::isSerializable):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::getContentFrameInfo):
(TestWebKitAPI::TEST(NodeInfo, Basic)):
(TestWebKitAPI::TEST(SerializedNode, Basic)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView objectByEvaluatingJavaScript:inFrame:inContentWorld:]):
(-[WKWebView objectByCallingAsyncFunction:withArguments:]):
(-[WKWebView objectByCallingAsyncFunction:withArguments:inFrame:inContentWorld:]):

Canonical link: <a href="https://commits.webkit.org/298211@main">https://commits.webkit.org/298211@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de1ec636e2ffb3a7bb3db9ddb44641ac1c2b5b06

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114628 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34374 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24837 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120794 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65350 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f467647f-0561-46b7-a79a-5e608f4c83ed) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87133 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42037 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/91747e97-dc52-4876-a202-030ad35f1871) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27884 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102941 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67521 "Found 2 new API test failures: /WPE/TestMultiprocess:/webkit/WebKitWebView/multiprocess-create-ready-close, /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27071 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21065 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64471 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97264 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21175 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123996 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41640 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31095 "Found 1 new test failure: fast/block/transparent-outline-with-and-without-border-radius.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95947 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42017 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99126 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95728 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40891 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18732 "") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37739 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18367 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41518 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47030 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41104 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44411 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42854 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -223,6 +224,15 @@ JavaScriptEvaluationResult::JavaScriptEvaluationResult(JSGlobalContextRef contex
 
 JSValueRef JavaScriptEvaluationResult::toJS(JSGlobalContextRef context, Value&& root)
 {
+    auto globalObjectForms = [] (auto context) {
```
What do you mean by "forms"??

achristensen07 replied:
In this case the JSGlobalObject*, JSDOMGlobalObject*, and Document* all basically point to the same thing.  They are different forms of the same global object.  I agree the name isn't great.  Maybe globalObjectPointers?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/de1ec636e2ffb3a7bb3db9ddb44641ac1c2b5b06. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48858

achristensen07 committed:
Add ability for _WKNodeInfo to be used as an argument with WKWebView.callAsyncJavaScript
https://bugs.webkit.org/show_bug.cgi?id=296836
rdar://157338015

Reviewed by Ryosuke Niwa.

If the node does not exist any more, I chose to pass JS undefined in as a parameter.
This makes _WKNodeInfo kind of like a WeakRef, which returns undefined if the node doesn't exist any more.
If the node exists in a different document, I decided to also return JS undefined.
Both of these choices are arbitrary and non-final at this point.  Still prototyping its use.

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toJS):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::isSerializable):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::getContentFrameInfo):
(TestWebKitAPI::TEST(NodeInfo, Basic)):
(TestWebKitAPI::TEST(SerializedNode, Basic)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView objectByEvaluatingJavaScript:inFrame:inContentWorld:]):
(-[WKWebView objectByCallingAsyncFunction:withArguments:]):
(-[WKWebView objectByCallingAsyncFunction:withArguments:inFrame:inContentWorld:]):

Canonical link: https://commits.webkit.org/298211@main

webkit-commit-queue commented:
Committed 298211@main (2dbebd6180a9): <https://commits.webkit.org/298211@main>

Reviewed commits have been landed. Closing PR #48858 and removing active labels.
